File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-6.cs
public class Solution {
    public int LargestRectangleArea(int[] heights) {
        int res = 0;
        Stack<int> stack = new Stack<int>();
        for (int i = 0; i <= heights.Length; i++) {
            while(stack.Count > 0 && (i == heights.Length || heights[stack.Peek()] >= heights[i])){
                int p = stack.Peek();
                stack.Pop();
                int h = heights[p];
                int w = 0;
                if(stack.Count == 0) w = i;
                else w = i - stack.Peek() - 1;
                if(h * w > res)
                    res = h * w; 
            }
            stack.Push(i);
        }
        return res;
    }
}

File: Data Structures & Algorithms/depth-of-binary-tree/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
        std::stack<std::pair<int, TreeNode*>> st;
        int res{};
        if (root) st.push({1, root});
        while (!st.empty()) {
            auto p{st.top()};
            st.pop();
            auto depth{p.first};
            auto* node{p.second};
            if (depth > res) res = depth;
            if (node->left) st.push({depth + 1, node->left});
            if (node->right) st.push({depth + 1, node->right});

        }
        return res;
    }
};
